sql/mysqld.cc

```diff
@@ -6071,38 +6105,198 @@ void handle_connections_sockets()
   int flags=0,retval;
   st_vio *vio_tmp;
   bool is_unix_sock;
-#ifdef HAVE_POLL
+#ifdef HAVE_SYSTEMD
+  int systemd_n, systemd_fd;
+  union sockaddr_union socketpeer;
+  socklen_t socketpeer_len;
+  char const *family;
+#ifndef HAVE_POLL
+  int  pfs_flags[MAX_LISTEN_SOCKETS];
+  // systemd wants these too
+  MYSQL_SOCKET  pfs_fds[MAX_LISTEN_SOCKETS]; // used for non-polling interfaces
+#endif
+#endif // HAVE_SYSTEMD
   int socket_count= 0;
-  struct pollfd fds[3]; // for ip_sock, unix_sock and extra_ip_sock
-  MYSQL_SOCKET  pfs_fds[3]; // for performance schema
-#define setup_fds(X)                    \
+  short fd_type[MAX_LISTEN_SOCKETS];
+#ifdef HAVE_POLL
+  struct pollfd fds[MAX_LISTEN_SOCKETS]; // for ip_sock, extra_ip_sock and unix_sock (could be any order for systemd)
```

Comment:
ok. will do the third option with a extra check that it isn't in the mariadb.socket file before opening it.